Feature: Implement `blockchain.transaction.id_from_pos` method

Implement `blockchain.transaction.id_from_pos` protocol method as described in the [specs here](https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html#blockchain-transaction-id-from-pos).

I would like to work on this issue @romanz

Use `txids[*tx_pos]` (since `tx_pos` is `&usize`):
```
error[E0277]: the type `[bitcoin::Txid]` cannot be indexed by `&usize`
   --> src/electrum.rs:420:18
    |
420 |         let tx = txids[tx_pos];
    |                  ^^^^^^^^^^^^^ slice indices are of type `usize` or ranges of `usize`
    |
    = help: the trait `SliceIndex<[bitcoin::Txid]>` is not implemented for `&usize`
    = note: required because of the requirements on the impl of `std::ops::Index<&usize>` for `Vec<bitcoin::Txid>`
```
https://github.com/romanz/electrs/actions/runs/4074792155/jobs/7020360062

--- a/src/electrum.rs
+++ b/src/electrum.rs
@@ -16,7 +16,7 @@ use crate::{
     cache::Cache,
     config::{Config, ELECTRS_VERSION},
     daemon::{self, extract_bitcoind_error, Daemon},
-    merkle::Proof,
+    merkle::{Proof, self},
     metrics::{self, Histogram, Metrics},
     signals::Signal,
     status::ScriptHashStatus,
@@ -401,6 +401,33 @@ impl Rpc {
         }
     }
 
+    fn transaction_from_pos(&self, (height, tx_pos, merkle): &(usize, usize, bool)) -> Result<Value> {
+        let chain = self.tracker.chain();
+        let blockhash = match chain.get_block_hash(*height) {
+            None => bail!("missing block at {}", height),
+            Some(blockhash) => blockhash,
+        };
+        let txids = self.daemon.get_block_txids(blockhash)?;
+        let tx = txids[tx_pos];
+        if *merkle {
+            match txids.iter().position(|current_txid| *current_txid == *txid) {
+                None => bail!("missing txid {} in block {}", txid, blockhash),
+                Some(position) => {
+                    let proof = Proof::create(&txids, position);
+                    Ok(json!({
+                        "tx_id": tx,
+                        "merkle": proof.to_hex(),
+                    }))
+                }
+            }
+        }
+        else {
+            Ok(json!({
+                "tx_id": tx,
+            }))
+        }
+    }
+
     fn get_fee_histogram(&self) -> Result<Value> {
         Ok(json!(self.tracker.fees_histogram()))
     }
@@ -535,6 +562,7 @@ impl Rpc {
                 Params::TransactionBroadcast(args) => self.transaction_broadcast(args),
                 Params::TransactionGet(args) => self.transaction_get(args),
                 Params::TransactionGetMerkle(args) => self.transaction_get_merkle(args),
+                Params::TransactionFromPosition(args) => self.transaction_from_pos(args),
                 Params::Version(args) => self.version(args),
             };
             call.response(result)
@@ -562,6 +590,7 @@ enum Params {
     ScriptHashSubscribe((ScriptHash,)),
     TransactionGet(TxGetArgs),
     TransactionGetMerkle((Txid, usize)),
+    TransactionFromPosition((usize, usize, bool)),
     Version((String, Version)),
 }
 
@@ -580,6 +609,7 @@ impl Params {
             "blockchain.transaction.broadcast" => Params::TransactionBroadcast(convert(params)?),
             "blockchain.transaction.get" => Params::TransactionGet(convert(params)?),
             "blockchain.transaction.get_merkle" => Params::TransactionGetMerkle(convert(params)?),
+            "blockchain.transaction.id_from_pos" => Params::TransactionFromPosition(convert(params)?),
             "mempool.get_fee_histogram" => Params::MempoolFeeHistogram,
             "server.banner" => Params::Banner,
             "server.donation_address" => Params::Donation,

--- a/src/electrum.rs
+++ b/src/electrum.rs
@@ -16,7 +16,7 @@ use crate::{
     cache::Cache,
     config::{Config, ELECTRS_VERSION},
     daemon::{self, extract_bitcoind_error, Daemon},
-    merkle::{Proof},
+    merkle::Proof,
     metrics::{self, Histogram, Metrics},
     signals::Signal,
     status::ScriptHashStatus,
@@ -401,14 +401,17 @@ impl Rpc {
         }
     }
 
-    fn transaction_from_pos(&self, (height, tx_pos, merkle): &(usize, usize, bool)) -> Result<Value> {
+    fn transaction_from_pos(
+        &self,
+        (height, tx_pos, merkle): &(usize, usize, bool),
+    ) -> Result<Value> {
         let chain = self.tracker.chain();
         let blockhash = match chain.get_block_hash(*height) {
             None => bail!("missing block at {}", height),
             Some(blockhash) => blockhash,
         };
         let txids = self.daemon.get_block_txids(blockhash)?;
-        let txid:Txid = txids[*tx_pos];
+        let txid: Txid = txids[*tx_pos];
         if *merkle {
             match txids.iter().position(|current_txid| *current_txid == txid) {
                 None => bail!("missing txid {} in block {}", txid, blockhash),
@@ -420,8 +423,7 @@ impl Rpc {
                     }))
                 }
             }
-        }
-        else {
+        } else {
             Ok(json!({
                 "tx_id": txid,
             }))
@@ -609,7 +611,9 @@ impl Params {
             "blockchain.transaction.broadcast" => Params::TransactionBroadcast(convert(params)?),
             "blockchain.transaction.get" => Params::TransactionGet(convert(params)?),
             "blockchain.transaction.get_merkle" => Params::TransactionGetMerkle(convert(params)?),
-            "blockchain.transaction.id_from_pos" => Params::TransactionFromPosition(convert(params)?),
+            "blockchain.transaction.id_from_pos" => {
+                Params::TransactionFromPosition(convert(params)?)
+            }
             "mempool.get_fee_histogram" => Params::MempoolFeeHistogram,
             "server.banner" => Params::Banner,
             "server.donation_address" => Params::Donation,

--- a/src/electrum.rs
+++ b/src/electrum.rs
@@ -16,7 +16,7 @@ use crate::{
     cache::Cache,
     config::{Config, ELECTRS_VERSION},
     daemon::{self, extract_bitcoind_error, Daemon},
-    merkle::{Proof, self},
+    merkle::{Proof},
     metrics::{self, Histogram, Metrics},
     signals::Signal,
     status::ScriptHashStatus,
@@ -408,14 +408,14 @@ impl Rpc {
             Some(blockhash) => blockhash,
         };
         let txids = self.daemon.get_block_txids(blockhash)?;
-        let tx = txids[tx_pos];
+        let txid:Txid = txids[*tx_pos];
         if *merkle {
-            match txids.iter().position(|current_txid| *current_txid == *txid) {
+            match txids.iter().position(|current_txid| *current_txid == txid) {
                 None => bail!("missing txid {} in block {}", txid, blockhash),
                 Some(position) => {
                     let proof = Proof::create(&txids, position);
                     Ok(json!({
-                        "tx_id": tx,
+                        "tx_id": txid,
                         "merkle": proof.to_hex(),
                     }))
                 }
@@ -423,7 +423,7 @@ impl Rpc {
         }
         else {
             Ok(json!({
-                "tx_id": tx,
+                "tx_id": txid,
             }))
         }
     }

--- a/src/electrum.rs
+++ b/src/electrum.rs
@@ -411,6 +411,9 @@ impl Rpc {
             Some(blockhash) => blockhash,
         };
         let txids = self.daemon.get_block_txids(blockhash)?;
+        if *tx_pos >= txids.len() {
+            bail!("invalid tx_pos {} in block at height {}", tx_pos, height);
+        }
         let txid: Txid = txids[*tx_pos];
         if *merkle {
             match txids.iter().position(|current_txid| *current_txid == txid) {

--- a/src/electrum.rs
+++ b/src/electrum.rs
@@ -403,33 +403,23 @@ impl Rpc {
 
     fn transaction_from_pos(
         &self,
-        (height, tx_pos, merkle): &(usize, usize, bool),
+        (height, tx_pos, merkle): (usize, usize, bool),
     ) -> Result<Value> {
         let chain = self.tracker.chain();
-        let blockhash = match chain.get_block_hash(*height) {
+        let blockhash = match chain.get_block_hash(height) {
             None => bail!("missing block at {}", height),
             Some(blockhash) => blockhash,
         };
         let txids = self.daemon.get_block_txids(blockhash)?;
-        if *tx_pos >= txids.len() {
+        if tx_pos >= txids.len() {
             bail!("invalid tx_pos {} in block at height {}", tx_pos, height);
         }
-        let txid: Txid = txids[*tx_pos];
-        if *merkle {
-            match txids.iter().position(|current_txid| *current_txid == txid) {
-                None => bail!("missing txid {} in block {}", txid, blockhash),
-                Some(position) => {
-                    let proof = Proof::create(&txids, position);
-                    Ok(json!({
-                        "tx_id": txid,
-                        "merkle": proof.to_hex(),
-                    }))
-                }
-            }
+        let txid: Txid = txids[tx_pos];
+        if merkle {
+            let proof = Proof::create(&txids, tx_pos);
+            Ok(json!({"tx_id": txid, "merkle": proof.to_hex()}))
         } else {
-            Ok(json!({
-                "tx_id": txid,
-            }))
+            Ok(json!({ "tx_id": txid }))
         }
     }
 
@@ -567,7 +557,7 @@ impl Rpc {
                 Params::TransactionBroadcast(args) => self.transaction_broadcast(args),
                 Params::TransactionGet(args) => self.transaction_get(args),
                 Params::TransactionGetMerkle(args) => self.transaction_get_merkle(args),
-                Params::TransactionFromPosition(args) => self.transaction_from_pos(args),
+                Params::TransactionFromPosition(args) => self.transaction_from_pos(*args),
                 Params::Version(args) => self.version(args),
             };
             call.response(result)

